Add: Система склонений предметов

Ohar opened the pull request:
## What Does This PR Do
Добавлена система склонений предметов (через свойство `name_ru`, процедуру `declent_ru` и константы падежей), которую можно применять для безболезненного использования называния предметов на русском языке в правильных падежах

Включает пример использования в сообщении о суициде челюстями жизни.

## Why It's Good For The Game

Система будет применяться во всех будущих переводах и послужит основой для нормального перевода названий предметов и одной из точек использования будущего переключателя языка.

## Changelog
:cl:
add: Добавлена система склонений предметов
/:cl:

Ohar committed:
Add: Добавлена система склонений предметов (через свойство `name_ru`, процедуру `declent_ru` и константы падежей), которую можно применять для безболезненного использования называния предметов на русском языке в правильных падежах

Vallat commented on code/game/atoms.dm:
```diff
@@ -1035,3 +1045,11 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
+// Процедура выбора правильного падежа для любого предмета,если у него указан словарь «name_ru», примерно такой:
+// name_ru = list(NOMINATIVE = "челюсти жизни", GENITIVE = "челюстей жизни", DATIVE = "челюстям жизни", ACCUSATIVE = "челюсти жизни", INSTRUMENTAL = "челюстями жизни", PREPOSITIONAL = "челюстях жизни")
+/atom/proc/declent_ru(case_id)
+	if(name_ru != null && length(name_ru))
```
`length()` уже проверяет на нулл-кейс
```suggestion
	if(length(name_ru))
```

Vallat commented on code/game/atoms.dm:
```diff
@@ -1035,3 +1045,11 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
+/atom/proc/declent_ru(case_id)
+	if(name_ru != null && length(name_ru))
+		return name_ru[case_id] || name
+	else
+		return name
```
```suggestion
	return name
```

Vallat commented on code/game/atoms.dm:
```diff
@@ -54,6 +62,8 @@
 	var/chat_color
 	/// A luminescence-shifted value of the last color calculated for chatmessage overlays
 	var/chat_color_darkened
+	/// Список склонений названия атома
+	var/list/name_ru
```
Рекомендую в комментарий добавить пример заполнения списка для корректной работы

Ohar replied:
```suggestion
	/// Список склонений названия атома. Пример заполнения в любом наследнике атома
	/// name_ru = list(NOMINATIVE = "челюсти жизни", GENITIVE = "челюстей жизни", DATIVE = "челюстям жизни", ACCUSATIVE = "челюсти жизни", INSTRUMENTAL = "челюстями жизни", PREPOSITIONAL = "челюстях жизни")
	var/list/name_ru
```

Ohar committed:
Update code/game/atoms.dm

Co-authored-by: Vallat <vox.crit2013@gmail.com>

Ohar committed:
Update code/game/atoms.dm

Co-authored-by: Vallat <vox.crit2013@gmail.com>

Ohar committed:
Update code/game/atoms.dm

name_ru: добавлен пример заполнения

Co-authored-by: Vallat <vox.crit2013@gmail.com>

Vallat commented on code/game/atoms.dm:
```diff
@@ -1,3 +1,11 @@
+#define DATIVE "dative" // Дательный: дать полный доступ кому? Клоуну и ассистухе
+#define ACCUSATIVE "accusative" // Винительный: обвинить кого? Клоуна и ассистуху
+#define INSTRUMENTAL "instrumental" // Творительный: возить по полу кем? Клоуном и ассистухой
+#define PREPOSITIONAL "prepositional" // Предложный: прохладная история о ком? О клоуне и об ассистухе
```
```suggestion
#define NOMINATIVE 1 // Именительный: кто это? Клоун и ассистуха
#define GENITIVE 2 // Родительный: откусить кусок от кого? От клоуна и ассистухи
#define DATIVE 3 // Дательный: дать полный доступ кому? Клоуну и ассистухе
#define ACCUSATIVE 4 // Винительный: обвинить кого? Клоуна и ассистуху
#define INSTRUMENTAL 5 // Творительный: возить по полу кем? Клоуном и ассистухой
#define PREPOSITIONAL 6 // Предложный: прохладная история о ком? О клоуне и об ассистухе
```

Ohar replied:
Уже влил сам

Ohar committed:
Падежи русского языка: использовать цифры вместо строчных констант

Vallat commented on code/game/atoms.dm:
```diff
@@ -1035,3 +1046,10 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
+
+// Процедура выбора правильного падежа для любого предмета,если у него указан словарь «name_ru», примерно такой:
+// name_ru = list(NOMINATIVE = "челюсти жизни", GENITIVE = "челюстей жизни", DATIVE = "челюстям жизни", ACCUSATIVE = "челюсти жизни", INSTRUMENTAL = "челюстями жизни", PREPOSITIONAL = "челюстях жизни")
+/atom/proc/declent_ru(case_id)
+	if(length(name_ru))
+		return name_ru[case_id] || name
+	return name
```
```suggestion
/atom/proc/declent_ru(case_id, list/ru_names_override)
	var/list/list_to_use = ru_names_override || name_ru
	if(length(list_to_use))
		return list_to_use[case_id] || name
	return name
```

Ohar replied:
Аналогично, уже сам написал это

Ohar committed:
/atom/proc/declent_ru: добавлен второй аргумент, чтобы можно было пробрасывать иной список
https://discord.com/channels/617003227182792704/734823601110515882/979060674065989642

Ohar committed:
/atom/var/name_ru: переименована для удобства чтения
https://discord.com/channels/617003227182792704/734823601110515882/979064409378914304

Bizzonium commented on code/game/objects/items/tools/crowbar.dm:
```diff
@@ -69,11 +69,12 @@
-	user.visible_message("<span class='suicide'>[user] is putting [user.p_their()] head in [src]. It looks like [user.p_theyre()] trying to commit suicide!</span>")
-	playsound(loc, 'sound/items/jaws_pry.ogg', 50, 1, -1)
+	user.visible_message("<span class='suicide'>[user] помеща[pluralize_ru(user.gender,'ет','ют')] свою голову между лезвиями [src.declent_ru(GENITIVE)]. Похоже, [genderize_ru(user.gender,'он','она','оно','они')] пыта[pluralize_ru(user.gender,'ется','ются')] использовать [src.declent_ru(ACCUSATIVE)] для самоубийства!</span>")
+    	playsound(loc, 'sound/items/jaws_pry.ogg', 50, 1, -1)
```
тут ошибка с отступом, советую компилить в след раз

Bizzonium commented on code/game/objects/items/tools/crowbar.dm:
```diff
@@ -69,11 +69,12 @@
 /obj/item/crowbar/power/suicide_act(mob/user)
-	user.visible_message("<span class='suicide'>[user] is putting [user.p_their()] head in [src]. It looks like [user.p_theyre()] trying to commit suicide!</span>")
-	playsound(loc, 'sound/items/jaws_pry.ogg', 50, 1, -1)
+	user.visible_message("<span class='suicide'>[user] помеща[pluralize_ru(user.gender,'ет','ют')] свою голову между лезвиями [src.declent_ru(GENITIVE)]. Похоже, [genderize_ru(user.gender,'он','она','оно','они')] пыта[pluralize_ru(user.gender,'ется','ются')] использовать [src.declent_ru(ACCUSATIVE)] для самоубийства!</span>")
```
Одинарные кавычки вызывают ошибки
